### PR TITLE
feat: metal vrf

### PR DESCRIPTION
**Pull request by antoninrykalsky**

### CODE REVIEW

File: plugins/modules/metal_vrf_info.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+from ansible_specdoc.objects import SpecField, FieldType, SpecReturnValue
+import traceback
+
+from ansible_collections.equinix.cloud.plugins.module_utils.equinix import (
+    EquinixModule,
+    getSpecDocMeta,
+)
+
+module_spec = dict(
+    name=SpecField(
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf_info.py]: Right, name filter does not work well. On the negative side, if a non-existent name is passed, all vrfs are returned. 

```

    - name: filter by name / get VRF info
      equinix.cloud.metal_vrf_info:
        name: "{{ test_prefix }}"
        project_id: "{{ test_project.id }}"
      register: vrf_list

    - name: filter by name / assert
      assert:
        that:
        - "vrf_list.resources|length == 1"

    - name: filter by name / get non existing VRF info
      equinix.cloud.metal_vrf_info:
        name: "nonexisting"
        project_id: "{{ test_project.id }}"
      register: vrf_list1

    - name: filter by name / debug
      debug:
        var: vrf_list1

    - name: filter by name / assert nonexiting
      assert:
        that:
        - "vrf_list1.resources|length == 0"
```

I have removed that block from `module_spec` 
```
    name=SpecField(
        type=FieldType.string,
        description=['Filter VRF on substring in name attribute.'],
    ),
```

File: plugins/modules/metal_vrf_info.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        description=['Filter VRF on substring in name attribute.'],
+    ),
+    project_id=SpecField(
+        type=FieldType.string,
+        description=['Project ID where the VRF will be deployed.'],
+    ),
+)
+
+specdoc_examples = ['''
+- name: Gather information about all resources in parent resource
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf_info.py]: replaced

File: plugins/modules/metal_vrf_info.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+module_spec = dict(
+    name=SpecField(
+        type=FieldType.string,
+        description=['Filter VRF on substring in name attribute.'],
+    ),
+    project_id=SpecField(
+        type=FieldType.string,
+        description=['Project ID where the VRF will be deployed.'],
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf_info.py]: replaced

File: plugins/modules/metal_vrf.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    \n}\n"
+  type: dict
+'''
+
+# End of generated documentation
+
+# This is a template for a new module. It is not meant to be used as is.
+# It is meant to be copied and modified to create a new module.
+# Replace all occurrences of "metal_resource" with the name of the new
+# module, for example "metal_vlan".
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: removed

File: plugins/modules/metal_vrf.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    : \"ansible-integration-test-vrf-6yww6pyz\",\n    \"project_id\": \"9934e474-04a1-46a3-842b-5f3dc0ed0eba\"\
+    \n}\n"
+  type: dict
+'''
+
+# End of generated documentation
+
+# This is a template for a new module. It is not meant to be used as is.
+# It is meant to be copied and modified to create a new module.
+# Replace all occurrences of "metal_resource" with the name of the new
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: removed

File: plugins/modules/metal_vrf.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    ,\n        \"id\": \"108b2cfb-246b-45e3-885a-bf3e82fce1a0\"\n    },\n    \"name\"\
+    : \"ansible-integration-test-vrf-6yww6pyz\",\n    \"project_id\": \"9934e474-04a1-46a3-842b-5f3dc0ed0eba\"\
+    \n}\n"
+  type: dict
+'''
+
+# End of generated documentation
+
+# This is a template for a new module. It is not meant to be used as is.
+# It is meant to be copied and modified to create a new module.
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: removed

File: plugins/modules/metal_vrf.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    : 65000,\n    \"metro\": {\n        \"href\": \"/metal/v1/locations/metros/108b2cfb-246b-45e3-885a-bf3e82fce1a0\"\
+    ,\n        \"id\": \"108b2cfb-246b-45e3-885a-bf3e82fce1a0\"\n    },\n    \"name\"\
+    : \"ansible-integration-test-vrf-6yww6pyz\",\n    \"project_id\": \"9934e474-04a1-46a3-842b-5f3dc0ed0eba\"\
+    \n}\n"
+  type: dict
+'''
+
+# End of generated documentation
+
+# This is a template for a new module. It is not meant to be used as is.
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: removed, thx

File: plugins/modules/metal_vrf.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+''']
+
+MUTABLE_ATTRIBUTES = [
+    k for k, v in module_spec.items() if v.editable
+]
+
+SPECDOC_META = getSpecDocMeta(
+    short_description='Use this resource to manage a VRF.',
+    description=(
+        'Create a VRF in your desired metro and project with any IP ranges that you want the VRF to route and forward.'
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: replaced, thx

File: tests/integration/targets/metal_vrf/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+          - "192.168.200.0/25"
+      register: test_vrf_2
+
+    - name: assert VRF created
+      assert:
+        that:
+          - "test_vrf.local_asn == 65000"
+          - "test_vrf_2.changed == false"
+
+    - name: get VRF info
```
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply tests/integration/targets/metal_vrf/tasks/main.yml:53]: Fixed. I've copied it from metal_vlan, but it contains that check.

File: plugins/modules/metal_vrf.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  name:
+    description:
+    - User-supplied name of the VRF, unique to the project.
+    required: false
+    type: str
+  project_id:
+    description:
+    - Project ID where the VRF will be deployed.
+    required: false
+    type: str
```
Comment (t0mk) [inline plugins/modules/metal_vrf.py:53]: There's more parameters to VRF:

https://github.com/equinix-labs/metal-python/blob/v0.3.0/equinix_metal/equinix_metal/models/vrf_create_input.py#L25
-   bgp_dynamic_neighbors_bfd_enabled
-   bgp_dynamic_neighbors_enabled
-   bgp_dynamic_neighbors_export_route_map

Some of the arguments of VRF are editable:
https://github.com/equinix-labs/metal-python/blob/v0.3.0/equinix_metal/equinix_metal/models/vrf_update_input.py#L28

They should be marked ` editable=True`

**Review by t0mk — CHANGES_REQUESTED** [review]

File: plugins/modules/metal_vrf.py (hunk 10 above)
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py:53]: Well, yeah, I was basing it on Terraform, on this 

https://registry.terraform.io/providers/equinix/equinix/latest/docs/resources/equinix_metal_vrf

there's nothing in there about those features. I'm going to have to look into this more in depth, it falls to me

```
pydantic.error_wrappers.ValidationError: 1 validation error for FindVrfs
id       
  field required (type=value_error.missing)
fatal: [testhost]: FAILED! => {
    "changed": false,                  
    "invocation": {
        "module_args": {                                                        
            "bgp_dynamic_neighbors_bfd_enabled": true,                                                                                                                            
            "bgp_dynamic_neighbors_enabled": true,                                                         
            "bgp_dynamic_neighbors_export_route_map": true,
            "description": "Test VRF with ASN 65000",
            "id": null,                                 
            "ip_ranges": [                                                                                                                                                                                                                                                   
                "192.168.100.0/25",
                "192.168.200.0/25"
            ],                        
            "local_asn": 65000,
            "metal_api_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "metal_api_url": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "metal_ua_prefix": "",               
            "metro": "am",
            "name": "ansible-integration-test-vrf-2r8qp35j",
            "project_id": null,              
            "state": "present" 
        }                                                
    },                                                               
    "msg": "Error in metal_vrf: 1 validation error for FindVrfs\nid\n  field required (type=value_error.missing)"
}                                 

```

Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py:53]:
```
fatal: [testhost]: FAILED! => {                                                                                                                                                   
    "changed": false,
    "invocation": {   
        "module_args": {                                      
            "bgp_dynamic_neighbors_bfd_enabled": false,
            "bgp_dynamic_neighbors_enabled": false,
            "bgp_dynamic_neighbors_export_route_map": false,                    
            "description": "Test VRF with ASN 65000",                                                                                                                             
            "id": null,                                                                                    
            "ip_ranges": [                                                      
                "192.168.100.0/25",                                                                                                                                               
                "192.168.200.0/25"                                                                         
            ],                                                                                                                                                                                                                                                               
            "local_asn": 65000,                     
            "metal_api_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "metal_api_url": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "metal_ua_prefix": "",    
            "metro": "am",
            "name": "ansible-integration-test-vrf-dsyhw88h",
            "project_id": "406e3b46-608a-4d4c-b2eb-a6dc30a9c668",
            "state": "present"                   
        }               
    },                                       
    "msg": "Error in metal_vrf: attribute 'bgp_dynamic_neighbors_bfd_enabled' (to map to 'bgp_dynamic_neighbors_bfd_enabled') not found in response_dict: {'bill': False, 'created_at': datetime.datetime(2023, 12, 13, 8, 37, 58, tzinfo=datetime.timezone.utc), 'created_b$
': {'href': '/metal/v1/users/ae8b343c-8800-44ff-a31f-edd1a2cbf86d', 'id': 'ae8b343c-8800-44ff-a31f-edd1a2cbf86d'}, 'description': 'Test VRF with ASN 65000', 'href': '/metal/v1/vrfs/b3a19195-aede-4f9e-bf21-3bd24d91f75a', 'id': 'b3a19195-aede-4f9e-bf21-3bd24d91f75a', 'i$
_ranges': ['192.168.100.0/25', '192.168.200.0/25'], 'local_asn': 65000, 'metro': {'href': '/metal/v1/locations/metros/108b2cfb-246b-45e3-885a-bf3e82fce1a0', 'id': '108b2cfb-246b-45e3-885a-bf3e82fce1a0'}, 'name': 'ansible-integration-test-vrf-dsyhw88h', 'project': {'hr$
f': '/metal/v1/projects/406e3b46-608a-4d4c-b2eb-a6dc30a9c668', 'id': '406e3b46-608a-4d4c-b2eb-a6dc30a9c668'}, 'tags': [], 'updated_at': datetime.datetime(2023, 12, 13, 8, 37, 58, tzinfo=datetime.timezone.utc), 'virtual_circuits': []}"
}

Comment (t0mk) [reply plugins/modules/metal_vrf.py:68]: Yes, the bgp flags are not present in response, let's skip.

**Review by t0mk — APPROVED** [review]

File: plugins/modules/metal_vrf.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    "invocation": {
+        "module_args": {
+            "description": null,
+            "id": "f4a7863c-fcbf-419c-802c-3c6d3ad9529e",
+            "ip_ranges": [
+                "192.168.100.0/25",
+                "192.168.200.0/25"
+            ],
+            "local_asn": 65000,
+            "metal_api_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
```
[the remark this answers is not in the document]
Comment (t0mk) [reply plugins/modules/metal_vrf.py:244]: We tend to keep the result sample simple - we only put here the dict from the `module_args`, without the `metal_*` keys. Please make the item in the result_sample list look just like that. Example in metal_project:

https://github.com/equinix-labs/ansible-collection-equinix/blob/main/plugins/modules/metal_project.py#L175

**Review by t0mk — CHANGES_REQUESTED** [review]

File: plugins/modules/metal_vrf.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py]: got it

File: plugins/modules/metal_vrf.py (hunk 10 above)
Comment (antoninrykalsky) [reply plugins/modules/metal_vrf.py:53]: I have removed it

**Review by t0mk — APPROVED** [review]

**Pull request merged**